File: Game Client/game/Custom.py
from .config import *
from .Ball import *
from .Button import *
from .Player import *
from .Wall import *
from .StartScreen import *
from .AI import *
from .Obstacle import *
from .WaitScreen import *

class CustomMenu:
	def __init__(self):
		self.size = [winWidth * 0.2, winHeight * 0.1]
		self.mod_size = [winWidth * 0.11, winHeight * 0.07]
		self.font = pg.font.Font(font, int(winHeight * 0.06))
		self.title_font = pg.font.Font(font, int(winHeight * 0.085))
		self.score = 10
		self.ai_nb = 0
		self.max_ai = 2
		self.down_buttons = [Button("BACK TO MENU", 25, winHeight - self.size[1] - 25, self.size[0], self.size[1], winHeight * 0.085),
				Button("START", winWidth - self.size[0] - 25, winHeight - self.size[1] - 25, self.size[0], self.size[1], winHeight * 0.085)]
		self.players_buttons = [Button("AI VS AI", winWidth / 5 - (self.mod_size[0] / 2), winHeight / 4 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
       			Button("1 VS 1", winWidth / 5 * 2 - (self.mod_size[0] / 2), winHeight / 4 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
          		Button("2 VS 2", winWidth / 5 * 3 - (self.mod_size[0] / 2), winHeight / 4 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
            	Button("1V1V1V1", winWidth / 5 * 4 - (self.mod_size[0] / 2), winHeight / 4 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06)]
		self.mod_buttons = [Button("LOCAL", winWidth / 5 - (self.mod_size[0] / 2), winHeight / 2 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
              	Button("ONLINE", winWidth / 5 * 2 - (self.mod_size[0] / 2), winHeight / 2 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
          		Button("BORDERLESS", winWidth / 5 * 3 - (self.mod_size[0] / 2), winHeight / 2 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
            	Button("OBSTACLE", winWidth / 5 * 4 - (self.mod_size[0] / 2), winHeight / 2 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06)]
		# self.param_buttons = [Button("-", winWidth / 2 + 125, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
		# 		Button("+", winWidth / 2 + 155, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
    	# 		Button("-", (winWidth / 5 * 4) + 140, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
		# 		Button("+", (winWidth / 5 * 4) + 170, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06)] # old pos
		self.param_buttons = [Button("-", winWidth / 3 + 125, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("+", winWidth / 3 + 155, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("-", winWidth / 3 * 2 + 140, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("+", winWidth / 3 * 2 + 170, winHeight / 4 * 3 - (winHeight * 0.03 / 2), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06)]
		self.players_buttons[1].highlight = True
		self.mod_buttons[0].highlight = True
	
	def draw(self, win):
		title = self.title_font.render("CUSTOM", True, (255, 255, 255))
		win.blit(title, (winWidth / 2 - (title.get_size()[0] / 2), winHeight * 0.1 - title.get_size()[1] * 0.45))
		score = self.font.render("MAX SCORE = " + str(self.score), True, (255, 255, 255))
		win.blit(score, (winWidth / 3 - (score.get_size()[0] / 2), winHeight / 4 * 3 - score.get_size()[1] * 0.45))
		# win.blit(score, (winWidth / 2 - (score.get_size()[0] / 2), winHeight / 4 * 3 - score.get_size()[1] * 0.45)) #center
		ai = self.font.render("AI OPPONENTS = " + str(self.ai_nb), True, (255, 255, 255))
		win.blit(ai, (winWidth / 3 * 2 - (ai.get_size()[0] / 2), winHeight / 4 * 3 - ai.get_size()[1] * 0.45))
		# win.blit(ai, (winWidth / 5 * 4 - (ai.get_size()[0] / 2), winHeight / 4 * 3 - ai.get_size()[1] * 0.45)) #right
		for button in self.down_buttons:
			button.draw(win)
		for button in self.players_buttons:
			button.draw(win)
		for button in self.mod_buttons:
			button.draw(win)
		for button in self.param_buttons:
			button.draw(win)

	async def click(self, core, mousePos):
		for button in self.down_buttons:
			if button.hitbox.collidepoint(mousePos):
				if button.name == "BACK TO MENU":
					core.state = "menu"
					core.mode = "none"
					core.max_score = 10
					break
				elif button.name == "START":
					await self.start(core)
					break

		for button in self.players_buttons:
			if button.hitbox.collidepoint(mousePos):
				if not button.highlight:
					for other in self.players_buttons:
						other.highlight = False
				button.highlight = not button.highlight
				self.max_ai = 2 if button.name == "AI VS AI" or button.name == "1 VS 1" else 4
				self.ai_nb = 0
				break

		for button in self.mod_buttons:
			if button.hitbox.collidepoint(mousePos):
				if button.name == "LOCAL" and not button.highlight:
					self.mod_buttons[1].highlight = False
				if button.name == "ONLINE" and not button.highlight:
					self.mod_buttons[0].highlight = False
				button.highlight = not button.highlight
				break

		for button in self.param_buttons:
			if button.hitbox.collidepoint(mousePos):
				if button == self.param_buttons[0] or button == self.param_buttons[1]:
					if button.name == "+":
						self.score += 1
					elif button.name == "-" and self.score > 0:
						self.score -= 1
				else:
					if button.name == "+" and self.ai_nb < self.max_ai:
						self.ai_nb += 1
					elif button.name == "-" and self.ai_nb > 0:
						self.ai_nb -= 1
				break

		if self.players_buttons[0].highlight:
			self.mod_buttons[0].highlight = True
			self.mod_buttons[1].highlight = False
			self.ai_nb = 2
		elif self.players_buttons[1].highlight and self.ai_nb > 0:
			self.mod_buttons[0].highlight = True
			self.mod_buttons[1].highlight = False
		elif self.players_buttons[3].highlight:
			self.mod_buttons[2].highlight = False
   
		if self.players_buttons[2].highlight or self.players_buttons[3].highlight:
			if self.ai_nb >= 3:
				self.mod_buttons[0].highlight = True
				self.mod_buttons[1].highlight = False


	async def start(self, core):
		if not self.validStart():
			return
		self.getMods(core.alias)
		core.max_score = self.score
		if "LOCAL" in self.mod_list:
			self.initPlayers(core)
			self.initWalls(core)
			core.ball = Ball(True if "BORDERLESS" in self.mod_list else False)
			core.state = "start"
			core.mode = "LOCAL"
			if "OBSTACLE" in self.mod_list:
				core.obstacle = Obstacle()
			msg = {"type" : 'custom', 'online' : 'false'}
			await core.GameHub.send(json.dumps(msg))
		elif "ONLINE" in self.mod_list:
			core.mode = "ONLINE"
			msg = {"type" : 'custom', 'online' : 'true', 'mods' : self.mod_list, 'score' : self.score, 'ai' : self.ai_nb, 'players' : self.players.__len__()}
			await core.GameHub.send(json.dumps(msg))
			response : dict = json.loads(await core.GameHub.recv())
			core.GameSocket = response['socket']
			core.id = response['pos']
			core.state = "waiting"
			core.online = True
			core.wait_screen = WaitScreen(response['ID'], core.id, self.players.__len__(), "CUSTOM")
		if "1V1V1V1" in self.mod_list:
			core.custom_mod = "1V1V1V1"

		core.start_screen = StartScreen('custom', core.online, True if "1V1V1V1" in self.mod_list else False, self.players.__len__())

	def getMods(self, alias = 'Player'):
		self.mod_list = []
		for button in self.mod_buttons:
			if button.highlight:
				self.mod_list.append(button.name)
		
		self.players = {}
		for button in self.players_buttons:
			if button.highlight:
				if button.name == "AI VS AI":
					self.players[1] = "AI"
					self.players[2] = "AI"
					break
				elif button.name == "1 VS 1":
					self.players[1] = alias + "1" if self.ai_nb < 2 else "AI"
					self.players[2] = alias + "2" if self.ai_nb < 1 else "AI"
					break
				elif button.name == "2 VS 2":
					self.players[1] = alias + "1" if self.ai_nb < 4 else "AI"
					self.players[2] = alias + "2" if self.ai_nb < 3 else "AI"
					self.players[3] = alias + "3" if self.ai_nb < 2 else "AI"
					self.players[4] = alias + "4" if self.ai_nb < 1 else "AI"
					break
				elif button.name == "1V1V1V1":
					self.mod_list.append(button.name)
					self.players[1] = alias + "1" if self.ai_nb < 4 else "AI"
					self.players[2] = alias + "2" if self.ai_nb < 3 else "AI"
					self.players[3] = alias + "3" if self.ai_nb < 2 else "AI"
					self.players[4] = alias + "4" if self.ai_nb < 1 else "AI"

      
	def initPlayers(self, core):
		core.players = []
		for key, name in self.players.items():
			core.players.append(Player(key, name, self.players.__len__(), True if "BORDERLESS" in self.mod_list else False, True if "1V1V1V1" in self.mod_list else False))
		for player in core.players:
			if player.name == "AI":
				core.ai.append(AI(player))


	def initWalls(self, core):
		if "BORDERLESS" in self.mod_list:
			core.walls = False
		elif "1V1V1V1" in self.mod_list:
			core.walls = [Wall("up", True), Wall("down", True), Wall("left", True), Wall("right", True)]
		else:
			core.walls = [Wall("up", False), Wall("down", False)]

	def validStart(self):
		selected = False
		for button in self.players_buttons:
			if button.highlight:
				selected = True
				break
		if not selected:
			return False
		selected = False
		for button in self.mod_buttons:
			if button.name == "LOCAL" and button.highlight:
				selected = True
				break
			if button.name == "ONLINE" and button.highlight:
				selected = True
				break
		if not selected:
			return False
		return True

File: srcs/django_code/front/js/game/Custom.js
import { canvas, ctx } from "./canvas.js";
import { Button } from "./Button.js";
import { Wall } from "./Wall.js";
import { Player } from "./Player.js";
import { Obstacle } from "./Obstacle.js";
import { Ball } from "./Ball.js";
import { StartScreen } from "./StartScreen.js";
import { AI } from "./AI.js";
import { Vec2 } from "./Vec2.js";
import { is_colliding } from "./Hitbox.js";

export class CustomMenu {
	constructor() {
		this.size = [canvas.width * 0.2, canvas.height * 0.1];
		this.mod_size = [canvas.width * 0.11, canvas.height * 0.07];
		this.score = 5;
		this.ai_nb = 0;
		this.max_ai = 2;
		this.down_buttons = [new Button("BACK TO MENU", this.size[0] * 0.1, canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1]),
				new Button("START", canvas.width - (this.size[0] * 1.1), canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1])];
		this.players_buttons = [new Button("AI VS AI", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
       			new Button("1 VS 1", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("2 VS 2", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("1V1V1V1", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.mod_buttons = [new Button("LOCAL", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
              	new Button("ONLINE", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("BORDERLESS", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("OBSTACLE", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.param_buttons = [new Button("-", canvas.width * 0.415, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.435, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.755, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.775, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03)];
		this.players_buttons[1].highlight = true;
		this.mod_buttons[0].highlight = true;
	}

	draw() {
		ctx.fillText("CUSTOM", canvas.width / 2, canvas.height * 0.1);
		for (let b of this.down_buttons)
			b.draw();
		ctx.font = Math.floor(canvas.height * 0.06) + "px pong-teko";
		ctx.fillText("MAX SCORE = " + this.score, canvas.width / 3, canvas.height / 4 * 3);
		ctx.fillText("AI OPPONENTS = " + this.ai_nb, canvas.width / 3 * 2, canvas.height / 4 * 3);
		for (let b of this.players_buttons)
			b.draw();
		for (let b of this.mod_buttons)
			b.draw();
		for (let b of this.param_buttons)
			b.draw();
		ctx.font = Math.floor(canvas.height * 0.085) + "px pong-teko";
	}

	click(core, mousePos) {
		const pos = new Vec2(mousePos[0], mousePos[1]);
		for (let b of this.down_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.size)) {
				if (b.name === "BACK TO MENU") {
					core.state = "menu";
					core.mode = "none";
					core.max_score = 5;
					core.custom_menu = false;
					break;
				}
				else if (b.name === "START") {
					this.start(core);
					break;
				}
			}
		}
		for (let b of this.players_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.mod_size)) {
				if (!b.highlight)
					for (let other of this.players_buttons)
						other.highlight = false;
				b.highlight = !b.highlight;
				this.max_ai = (b.name === "AI VS AI" || b.name === "1 VS 1") ? 2 : 4;
				this.ai_nb = 0;
				break;
			}
		}
		for (let b of this.mod_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.mod_size)) {
				if (b.name === "LOCAL" && !b.highlight)
					this.mod_buttons[1].highlight = false;
				else if (b.name === "ONLINE" && !b.highlight)
					this.mod_buttons[0].highlight = false;
				b.highlight = !b.highlight;
				break;
			}
		}
		for (let b of this.param_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, [b.width, b.height])) {
				if (b === this.param_buttons[0] || b === this.param_buttons[1]) {
					if (b.name === "-" && this.score > 0)
						this.score -= 1;
					else if (b.name === "+")
						this.score += 1;
				}
				else {
					if (b.name === "-" && this.ai_nb > 0)
						this.ai_nb -= 1;
					else if (b.name === "+" && this.ai_nb < this.max_ai)
						this.ai_nb += 1;
				}
				break;
			}
		}

		if (this.players_buttons[0].highlight) {
			this.ai_nb = 2;
			this.mod_buttons[0].highlight = true;
			this.mod_buttons[1].highlight = false;
		}
		else if (this.players_buttons[1].highlight && this.ai_nb > 0) {
			this.mod_buttons[0].highlight = true;
			this.mod_buttons[1].highlight = false;
		}
		else if (this.players_buttons[3].highlight)
			this.mod_buttons[2].highlight = false;

		if (this.players_buttons[2].highlight || this.players_buttons[3].highlight) {
			if (this.ai_nb >= 3) {
				this.mod_buttons[0].highlight = true;
				this.mod_buttons[1].highlight = false;
			}
		}
	}

	start(core) {
		if (!this.validStart())
			return;
		this.getMods(core.alias);
		core.max_score = this.score;
		core.customs = this.mod_list;
		if (this.mod_list.includes("LOCAL")) {
			this.initPlayers(core);
			this.initWalls(core);
			core.ball = new Ball((this.mod_list.includes("BORDERLESS")) ? true : false);
			core.state = "start";
			core.mode = "LOCAL";
			if (this.mod_list.includes("OBSTACLE"))
				core.obstacle = new Obstacle();
			const msg = {"type" : 'custom', 'online' : 'false'};
			core.GameHub.send(JSON.stringify(msg));
		}
		else if (this.mod_list.includes("ONLINE")) {
			const msg = {"type" : 'custom', 'online' : 'true', 'mods' : this.mod_list, 'score' : this.score, 'ai' : this.ai_nb, 'players' : Object.keys(this.players).length};
			core.GameHub.send(JSON.stringify(msg));
			core.online = true;
		}
		if (this.mod_list.includes("1V1V1V1"))
			core.square = true;

		core.start_screen = new StartScreen("custom", core.online, core.square, Object.keys(this.players).length);
	}

	getMods(alias = "Player") {
		this.mod_list = [];
		for (let b of this.mod_buttons) {
			if (b.highlight)
				this.mod_list.push(b.name);
		}

		this.players = {};
		for (let b of this.players_buttons) {
			if (b.highlight) {
				if (b.name === "AI VS AI") {
					this.players[1] = "AI";
					this.players[2] = "AI";
				}
				else if (b.name === "1 VS 1") {
					this.players[1] = (this.ai_nb < 2) ? alias + "1" : "AI";
					this.players[2] = (this.ai_nb < 1) ? alias + "2" : "AI";
				}
				else if (b.name === "2 VS 2") {
					this.players[1] = (this.ai_nb < 4) ? alias + "1" : "AI";
					this.players[2] = (this.ai_nb < 3) ? alias + "2" : "AI";
					this.players[3] = (this.ai_nb < 2) ? alias + "3" : "AI";
					this.players[4] = (this.ai_nb < 1) ? alias + "4" : "AI";
				}
				else if (b.name === "1V1V1V1") {
					this.mod_list.push("1V1V1V1");
					this.players[1] = (this.ai_nb < 4) ? alias + "1" : "AI";
					this.players[2] = (this.ai_nb < 3) ? alias + "2" : "AI";
					this.players[3] = (this.ai_nb < 2) ? alias + "3" : "AI";
					this.players[4] = (this.ai_nb < 1) ? alias + "4" : "AI";
				}
				break;
			}
		}
	}

	initPlayers(core) {
		core.players = [];
		for (const key in this.players)
			core.players.push(new Player(parseInt(key), this.players[key], Object.keys(this.players).length, this.mod_list.includes("BORDERLESS"), this.mod_list.includes("1V1V1V1")));
		for (let p of core.players)
			if (p.name === "AI")
				core.ai.push(new AI(p));
	}

	initWalls(core) {
		if (this.mod_list.includes("BORDERLESS"))
			core.walls = false;
		else if (this.mod_list.includes("1V1V1V1"))
			core.walls = [new Wall("up", true), new Wall("down", true), new Wall("left", true), new Wall("right", true)];
		else
			core.walls = [new Wall("up", false), new Wall("down", false)];
	}

	validStart() {
		let selected = false;
		for (let b of this.players_buttons) {
			if (b.highlight) {
				selected = true;
				break;
			}
		}
		if (!selected)
			return false;
		selected = false;
		for (let b of this.mod_buttons) {
			if (b.highlight && b.name === "LOCAL") {
				selected = true;
				break;
			}
			if (b.highlight && b.name === "ONLINE") {
				selected = true;
				break;
			}
		}
		return selected;
	}

	responsive() {
		this.size = [canvas.width * 0.2, canvas.height * 0.1];
		this.mod_size = [canvas.width * 0.11, canvas.height * 0.07];

		this.down_buttons = [new Button("BACK TO MENU", this.size[0] * 0.1, canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1]),
				new Button("START", canvas.width - (this.size[0] * 1.1), canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1])];
		this.players_buttons = [new Button("AI VS AI", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
       			new Button("1 VS 1", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("2 VS 2", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("1V1V1V1", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 4 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.mod_buttons = [new Button("LOCAL", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
              	new Button("ONLINE", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("BORDERLESS", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("OBSTACLE", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 2 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.param_buttons = [new Button("-", canvas.width * 0.415, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.435, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.755, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.775, canvas.height / 4 * 3 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03)];
		this.players_buttons[1].highlight = true;
		this.mod_buttons[0].highlight = true;
	}
}